Mathlib/LinearAlgebra/RootSystem/Finite/CanonicalBilinear.lean

```diff
@@ -0,0 +1,139 @@
+/-
+Copyright (c) 2024 Scott Carnahan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Carnahan
+-/
+import Mathlib.LinearAlgebra.RootSystem.Defs
+import Mathlib.Algebra.Ring.SumsOfSquares
+
+/-!
+# The canonical bilinear form on a finite root pairing
+Given a finite root pairing, we define a canonical map from weight space to coweight space, and the
+corresponding bilinear form. This form is symmetric and Weyl-invariant, and if the base ring is
+linearly ordered, then the form is root-positive, positive-semidefinite on the weight space, and
+positive-definite on the span of roots.
+From these facts, it is easy to show that Coxeter weights in a finite root pairing are bounded
+above by 4. Thus, the pairings of roots and coroots in a crystallographic root pairing are
+restricted to a small finite set of possibilities.
+Another application is to the faithfulness of the Weyl group action on roots, and finiteness of the
+Weyl group.
+
+## Main definitions:
+ * `Polarization`: A distinguished linear map from the weight space to the coweight space.
+ * `CanonicalBilinear` : The bilinear form on weight space corresponding to `Polarization`.
+
+## References:
+ * SGAIII Exp. XXI
+ * Bourbaki, Lie groups and Lie algebras
+
+## Main results:
+ * `polarization_self_sum_of_squares` : The inner product of any weight vector is a sum of squares.
+ * `canonicalBilinear_reflection_reflection_apply` : `CanonicalBilinear` is invariant with respect
+   to reflections.
+ * `canonicalBilinear_self_smul_coroot`: The inner product of a root with itself times the
+   corresponding coroot is equal to two times Polarization applied to the root.
+
+## TODO (possibly in other files)
+ * Positivity and nondegeneracy
+ * Weyl-invariance
+ * Faithfulness of Weyl group action, and finiteness of Weyl group, for finite root systems.
+ * Relation to Coxeter weight.  In particular, positivity constraints for finite root pairings mean
+  we restrict to weights between 0 and 4.
+-/
+
+open Set Function
+open Module hiding reflection
+
+noncomputable section
+
+variable {ι R M N : Type*}
+
+lemma isSumSq_of_sum_of_squares [Mul R] [AddCommMonoid R] (s : Finset ι) (f : ι → R) :
+    IsSumSq (∑ i ∈ s, f i * f i) := by
+  induction s using Finset.cons_induction with
+  | empty =>
+    simpa only [Finset.sum_empty] using IsSumSq.zero
+  | cons i s his h =>
+    simp only [Finset.sum_cons]
+    exact IsSumSq.sq_add (f i) (∑ i ∈ s, f i * f i) h
+--#find_home! isSumSq_of_sum_of_squares --returns this file
```

Comment:
I think there's a very good case for just adding:
```
import Mathlib.Algebra.BigOperators.Group.Finset
```
to `Mathlib.Algebra.Ring.SumsOfSquares` and then placing this lemma there.

I think it should be named `isSumSq_sum_mul_self`.